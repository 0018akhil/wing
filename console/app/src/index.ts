import {
  createConsoleServer,
  CreateConsoleServerOptions,
} from "@wingconsole/server";
import express from "express";

export interface CreateConsoleAppOptions {
  wingfile: string;
  log?: CreateConsoleServerOptions["log"];
  updater?: CreateConsoleServerOptions["updater"];
  config?: CreateConsoleServerOptions["config"];
  requestedPort?: number;
  hostUtils?: CreateConsoleServerOptions["hostUtils"];
  onExpressCreated?: CreateConsoleServerOptions["onExpressCreated"];
}

export const createConsoleApp = async (options: CreateConsoleAppOptions) => {
  const server = await createConsoleServer({
    ...options,
    onExpressCreated(app) {
      // const files =
      //   process.env.NODE_ENV === "production"
      //     ? `${__dirname}/vite`
      //     : `${__dirname}/../dist/vite`;
      // app.use(express.static(files));
      app.use(express.static(`${__dirname}/vite`));
      options.onExpressCreated?.(app);
    },
    log: options.log ?? {
      info() {},
      error: console.error,
      verbose() {},
    },
    config: options.config ?? {
      addEventListener(event, listener) {},
      removeEventListener(event, listener) {},
      get(key) {
        return undefined as any;
      },
      set(key, value) {},
    },
  });
  return server;
};
