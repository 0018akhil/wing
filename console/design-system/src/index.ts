export * from "./toolbar.js";
export * from "./toolbar-button.js";
export * from "./tree-item.js";
export * from "./tree-view.js";
