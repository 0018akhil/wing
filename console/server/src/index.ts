import { inferRouterInputs } from "@trpc/server";
import Emittery from "emittery";

import { ConsoleLogger, createConsoleLogger } from "./consoleLogger.js";
import { createExpressServer } from "./expressServer.js";
import { Router } from "./router/index.js";
import { State } from "./types.js";
import { Updater } from "./updater.js";
import { createCompiler } from "./utils/compiler.js";
import { LogInterface } from "./utils/LogInterface.js";
import { createSimulator } from "./utils/simulator.js";

export type { LogEntry, LogLevel } from "./consoleLogger.js";
export type { ExplorerItem } from "./router/app.js";
export type { State } from "./types.js";
export type { WingSimulatorSchema, BaseResourceSchema } from "./wingsdk.js";
export type { Updater, UpdaterStatus } from "./updater.js";
export type { Router } from "./router/index.js";

type RouteNames = keyof inferRouterInputs<Router> | undefined;

export interface CreateConsoleServerOptions {
  wingfile: string;
  log: LogInterface;
  updater?: Updater;
  requestedPort?: number;
}

export const createConsoleServer = async ({
  wingfile,
  log,
  updater,
  requestedPort,
}: CreateConsoleServerOptions) => {
  const emitter = new Emittery<{
    invalidateQuery: RouteNames;
  }>();

  const invalidateQuery = async (query: RouteNames) => {
    await emitter.emit("invalidateQuery", query);
  };

  const invalidateUpdaterStatus = async () => {
    await invalidateQuery("updater.currentStatus");
  };
  updater?.addEventListener("status-change", invalidateUpdaterStatus);

  const consoleLogger: ConsoleLogger = createConsoleLogger({
    onLog: (level, message) => {
      invalidateQuery("app.logs");
    },
    log,
  });

  const compiler = createCompiler(wingfile);
  const simulator = createSimulator();
  compiler.on("compiled", ({ simfile }) => {
    simulator.start(simfile);
  });

  let lastErrorMessage = "";
  let appState: State = "compiling";
  compiler.on("compiling", () => {
    lastErrorMessage = "";
    appState = "compiling";
    invalidateQuery("app.state");
    invalidateQuery("app.error");
  });
  compiler.on("error", (error) => {
    lastErrorMessage = error.message;
    appState = "error";
    invalidateQuery("app.state");
    invalidateQuery("app.error");
  });
  simulator.on("starting", () => {
    appState = "loadingSimulator";
    invalidateQuery("app.state");
  });
  simulator.on("started", () => {
    appState = "success";
    invalidateQuery(undefined);
  });
  simulator.on("error", (error) => {
    lastErrorMessage = error.message;
    appState = "error";
    invalidateQuery("app.state");
    invalidateQuery("app.error");
  });

  // Create the express server and router for the simulator. Start
  // listening but don't wait for it, yet.
  log.info("Starting the dev server...");

  const { server, port } = await createExpressServer({
    consoleLogger,
    simulatorInstance() {
      return simulator.instance();
    },
    errorMessage() {
      return lastErrorMessage;
    },
    emitter: emitter as Emittery<{ invalidateQuery: string | undefined }>,
    log,
    updater,
    requestedPort,
    appState() {
      return appState;
    },
  });

  simulator.instance().then(async (simulatorInstance) => {
    simulatorInstance.onTrace({
      callback(event) {
        // TODO: Refactor the whole logs and events so we support all of the fields that the simulator uses.
        const message = `${
          event.data.message ?? JSON.stringify(event.data, undefined, 2)
        }`;
        if (event.type === "log") {
          consoleLogger.log(message, "simulator", {
            sourceType: event.sourceType,
            sourcePath: event.sourcePath,
          });
        } else {
          consoleLogger.verbose(message, "simulator", {
            sourceType: event.sourceType,
            sourcePath: event.sourcePath,
          });
        }
        if (event.data.status === "failure") {
          consoleLogger.error(event.data.error.message, "user", {
            sourceType: event.sourceType,
            sourcePath: event.sourcePath,
          });
        }

        if (
          event.sourceType === "wingsdk.cloud.Queue" &&
          // TODO: Change implementation after https://github.com/winglang/wing/issues/1713 is done
          event.data.message?.includes("Sending messages")
        ) {
          invalidateQuery("queue.approxSize");
        }
      },
    });
  });

  const close = async () => {
    try {
      updater?.removeEventListener("status-change", invalidateUpdaterStatus);
      await Promise.allSettled([
        server.close(),
        compiler.stop(),
        simulator.stop(),
      ]);
    } catch (error) {
      log.error(error);
    }
  };

  return {
    port,
    close,
  };
};
