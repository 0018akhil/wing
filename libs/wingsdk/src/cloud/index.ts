export * from "./bucket";
export * from "./function";
export * from "./queue";
export * from "./resource";
export * from "./shared";
