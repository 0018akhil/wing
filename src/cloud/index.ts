export * from "./bucket";
export * from "./function";
export * from "./resource";
export * from "./shared";
