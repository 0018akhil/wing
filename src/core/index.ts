export * from "./duration";
export * from "./process";
export * from "./app";
