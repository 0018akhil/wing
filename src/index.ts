export * as cloud from './cloud';
export * as core from './core';