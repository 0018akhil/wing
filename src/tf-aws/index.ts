export * from "./synth";
export * from "./bucket";
export * from "./factory";
export * from "./function";
