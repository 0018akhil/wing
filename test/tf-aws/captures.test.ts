import { Polycons } from "@monadahq/polycons";
import * as cdktf from "cdktf";
import * as cloud from "../../src/cloud";
import * as core from "../../src/core";
import * as tfaws from "../../src/tf-aws";
import { tfResourcesOf, tfSanitize } from "../util";

test("function captures some primitives", () => {
  const output = cdktf.Testing.synthScope((scope) => {
    const factory = new tfaws.PolyconFactory();
    Polycons.register(scope, factory);

    const inflight = new core.Inflight({
      code: core.NodeJsCode.fromInline(
        `async function $proc($cap, event) {
          console.log("Hello, " + $cap.string);
          console.log("My favorite number is " + $cap.num);
          console.log("The order is " + JSON.stringify($cap.order));
        }`
      ),
      entrypoint: "$proc",
      captures: {
        num: {
          obj: 5,
        },
        string: {
          obj: "world",
        },
        map: {
          obj: true,
        },
      },
    });
    const fn = new cloud.Function(scope, "Function", inflight);

    const code = core.Testing.inspectPrebundledCode(fn);
    expect(code.text).toMatchSnapshot();
  });

  expect(tfResourcesOf(output)).toEqual([
    "aws_iam_role",
    "aws_iam_role_policy",
    "aws_iam_role_policy_attachment",
    "aws_lambda_function",
    "aws_s3_bucket",
    "aws_s3_object",
  ]);
  expect(output).toMatchSnapshot();
});

test("function captures a bucket", () => {
  const output = cdktf.Testing.synthScope((scope) => {
    const factory = new tfaws.PolyconFactory();
    Polycons.register(scope, factory);

    const bucket = new cloud.Bucket(scope, "Bucket");
    const inflight = new core.Inflight({
      code: core.NodeJsCode.fromInline(
        `async function $proc($cap, event) {
          console.log("Hello, " + event.name);
          await $cap.bucket.put("hello.txt", JSON.stringify(event));
        }`
      ),
      entrypoint: "$proc",
      captures: {
        bucket: {
          obj: bucket,
          methods: ["put"],
        },
      },
    });
    const fn = new cloud.Function(scope, "Function", inflight);

    const code = core.Testing.inspectPrebundledCode(fn);
    expect(code.text).toMatchSnapshot();
  });

  expect(tfResourcesOf(output)).toEqual([
    "aws_iam_role",
    "aws_iam_role_policy",
    "aws_iam_role_policy_attachment",
    "aws_lambda_function",
    "aws_s3_bucket",
    "aws_s3_bucket_public_access_block",
    "aws_s3_bucket_server_side_encryption_configuration",
    "aws_s3_object",
  ]);
  expect(tfSanitize(output)).toMatchSnapshot();
});

test("two functions reusing the same inflight", () => {
  const output = cdktf.Testing.synthScope((scope) => {
    const factory = new tfaws.PolyconFactory();
    Polycons.register(scope, factory);

    const bucket = new cloud.Bucket(scope, "Bucket");
    const inflight = new core.Inflight({
      code: core.NodeJsCode.fromInline(
        `async function $proc($cap, event) {
          console.log("Hello, " + event.name);
          await $cap.bucket.put("hello.txt", JSON.stringify(event));
        }`
      ),
      entrypoint: "$proc",
      captures: {
        bucket: {
          obj: bucket,
          methods: ["put"],
        },
      },
    });
    const fn1 = new cloud.Function(scope, "Function1", inflight);
    const fn2 = new cloud.Function(scope, "Function2", inflight);

    expect(core.Testing.inspectPrebundledCode(fn1).text).toMatchSnapshot();
    expect(core.Testing.inspectPrebundledCode(fn2).text).toMatchSnapshot();
  });

  expect(tfResourcesOf(output)).toEqual([
    "aws_iam_role",
    "aws_iam_role_policy",
    "aws_iam_role_policy_attachment",
    "aws_lambda_function",
    "aws_s3_bucket",
    "aws_s3_bucket_public_access_block",
    "aws_s3_bucket_server_side_encryption_configuration",
    "aws_s3_object",
  ]);
  expect(tfSanitize(output)).toMatchSnapshot();
});
