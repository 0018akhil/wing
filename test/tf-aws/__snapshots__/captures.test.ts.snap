// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function captures a bucket 1`] = `
"{
  \\"resource\\": {
    \\"aws_iam_role\\": {
      \\"Function_IamRole_678BE84C\\": {
        \\"assume_role_policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":\\\\\\"sts:AssumeRole\\\\\\",\\\\\\"Principal\\\\\\":{\\\\\\"Service\\\\\\":\\\\\\"lambda.amazonaws.com\\\\\\"},\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\"
      }
    },
    \\"aws_iam_role_policy\\": {
      \\"Function_IamRolePolicy_E3B26607\\": {
        \\"policy\\": \\"{\\\\\\"Version\\\\\\":\\\\\\"2012-10-17\\\\\\",\\\\\\"Statement\\\\\\":[{\\\\\\"Action\\\\\\":[\\\\\\"s3:PutObject*\\\\\\",\\\\\\"s3:Abort*\\\\\\"],\\\\\\"Resource\\\\\\":[\\\\\\"\${aws_s3_bucket.Bucket.arn}\\\\\\",\\\\\\"\${aws_s3_bucket.Bucket.arn}/*\\\\\\"],\\\\\\"Effect\\\\\\":\\\\\\"Allow\\\\\\"}]}\\",
        \\"role\\": \\"\${aws_iam_role.Function_IamRole_678BE84C.name}\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"Function_IamRolePolicyAttachment_CACE1358\\": {
        \\"policy_arn\\": \\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\",
        \\"role\\": \\"\${aws_iam_role.Function_IamRole_678BE84C.name}\\"
      }
    },
    \\"aws_lambda_function\\": {
      \\"Function_LambdaFunction_27600CF9\\": {
        \\"environment\\": {
          \\"variables\\": {
            \\"BUCKET_NAME__Bucket\\": \\"\${aws_s3_bucket.Bucket.bucket}\\",
            \\"WING_FUNCTION_NAME\\": \\"Function\\"
          }
        },
        \\"function_name\\": \\"Function\\",
        \\"handler\\": \\"index.handler\\",
        \\"role\\": \\"\${aws_iam_role.Function_IamRole_678BE84C.arn}\\",
        \\"runtime\\": \\"nodejs16.x\\",
        \\"s3_bucket\\": \\"\${aws_s3_bucket.Function_Bucket_CC3697BD.bucket}\\",
        \\"s3_key\\": \\"\${aws_s3_object.Function_S3Object_C62A0C2D.key}\\"
      }
    },
    \\"aws_s3_bucket\\": {
      \\"Bucket\\": {
      },
      \\"Function_Bucket_CC3697BD\\": {
      }
    },
    \\"aws_s3_bucket_public_access_block\\": {
      \\"Bucket_PublicAccessBlock_A34F3B5C\\": {
        \\"block_public_acls\\": true,
        \\"block_public_policy\\": true,
        \\"bucket\\": \\"\${aws_s3_bucket.Bucket.bucket}\\",
        \\"ignore_public_acls\\": true,
        \\"restrict_public_buckets\\": true
      }
    },
    \\"aws_s3_bucket_server_side_encryption_configuration\\": {
      \\"Bucket_Encryption_016FDA0C\\": {
        \\"bucket\\": \\"\${aws_s3_bucket.Bucket.bucket}\\",
        \\"rule\\": [
          {
            \\"apply_server_side_encryption_by_default\\": {
              \\"sse_algorithm\\": \\"AES256\\"
            }
          }
        ]
      }
    },
    \\"aws_s3_object\\": {
      \\"Function_S3Object_C62A0C2D\\": {
        \\"bucket\\": \\"\${aws_s3_bucket.Function_Bucket_CC3697BD.bucket}\\",
        \\"key\\": \\"c83ba82ab737b318be3084bef5b7996129facf99b8.98fc77144a02c0c0f1283d26c18e6d53.zip\\",
        \\"source\\": \\"assets/Function_Asset_212D1EED/98FC77144A02C0C0F1283D26C18E6D53/archive.zip\\"
      }
    }
  }
}"
`;
